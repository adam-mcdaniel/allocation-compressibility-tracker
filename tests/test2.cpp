#include <stdio.h>
#include <stdlib.h>
#include <iostream>
#include <pthread.h>
#include <unistd.h>
#include <thread>
#include <chrono>

#define MEMORY_SIZE 1024  // Size of the allocated memory

void* threadFunction(void* arg) {
    char* memory = (char*)malloc(MEMORY_SIZE);  // Allocate memory

    if (memory == NULL) {
        fprintf(stderr, "Memory allocation failed.\n");
        pthread_exit(NULL);
    }

    // Loop until 10 seconds have elapsed
    auto start = std::chrono::high_resolution_clock::now();

    while (true) {
        if (std::chrono::duration_cast<std::chrono::seconds>(
                std::chrono::high_resolution_clock::now() - start)
                .count() > 10) {
            break;
        }
        // Write to the allocated memory
        // Here, we are simply writing 'A' to the memory repeatedly
        for (int i = 0; i < MEMORY_SIZE; i++) {
            int *x = (int*)malloc(10 * i);
            std::this_thread::sleep_for(std::chrono::milliseconds(50));
            memory[i] = 'A';
            free(x);
        }
    }

    free(memory);
    pthread_exit(NULL);
}

int main() {
    pthread_t threads[2];

    for (int i = 0; i < 2; i++) {
        int result = pthread_create(&threads[i], NULL, threadFunction, NULL);
        if (result != 0) {
            fprintf(stderr, "Thread creation failed.\n");
            return 1;
        }
    }

    // Wait for the threads to finish (never executed)
    for (int i = 0; i < 2; i++) {
        pthread_join(threads[i], NULL);
    }

    return 0;
}
