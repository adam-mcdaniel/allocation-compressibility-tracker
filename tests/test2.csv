Interval # (10000 ms), Bucket Size, Number of Allocations, Buckets's Compression Efficiency (compressed/uncompressed), Bucket's Uncompressed Portion of Heap, Bucket's Compressed Portion of Heap, Bucket's Uncompressed Size, Bucket's Compressed Size, Total Uncompressed Heap Size, Total Compressed Heap Size, 
1,16-32,1,1.18519,0.00332226,0.062645,32,27,9632,431,
1,64-128,1,1.24272,0.013289,0.238979,128,103,9632,431,
1,128-256,3,5.55372,0.0697674,0.280742,672,121,9632,431,
1,256-512,2,10.3226,0.0664452,0.143852,640,62,9632,431,
1,512-1024,2,51.2,0.212625,0.0928074,2048,40,9632,431,
1,1024-2048,1,87.6522,0.209302,0.0533643,2016,23,9632,431,
1,2048-4096,1,74.4727,0.425249,0.12761,4096,55,9632,431,
2,16-32,2,1.18519,0.00301205,0.06,64,54,21248,900,
2,64-128,2,1.24272,0.0120482,0.228889,256,206,21248,900,
2,128-256,6,5.31225,0.063253,0.281111,1344,253,21248,900,
2,256-512,4,10.3226,0.060241,0.137778,1280,124,21248,900,
2,512-1024,4,51.2,0.192771,0.0888889,4096,80,21248,900,
2,1024-2048,1,87.6522,0.0948795,0.0255556,2016,23,21248,900,
2,2048-4096,3,76.2,0.573795,0.177778,12192,160,21248,900,
3,16-32,3,1.18519,0.00332594,0.0604478,96,81,28864,1340,
3,64-128,3,1.24272,0.0133038,0.230597,384,309,28864,1340,
3,128-256,9,5.25,0.0698448,0.286567,2016,384,28864,1340,
3,256-512,6,10.3226,0.0665188,0.138806,1920,186,28864,1340,
3,512-1024,6,51.2,0.21286,0.0895522,6144,120,28864,1340,
3,1024-2048,1,87.6522,0.0698448,0.0171642,2016,23,28864,1340,
3,2048-4096,4,68.7257,0.564302,0.176866,16288,237,28864,1340,
4,16-32,4,1.18519,0.00350877,0.0607083,128,108,36480,1779,
4,64-128,4,1.24272,0.0140351,0.231591,512,412,36480,1779,
4,128-256,12,5.23977,0.0736842,0.288364,2688,513,36480,1779,
4,256-512,8,10.3226,0.0701754,0.139404,2560,248,36480,1779,
4,512-1024,8,51.2,0.224561,0.0899382,8192,160,36480,1779,
4,1024-2048,1,87.6522,0.0552632,0.0129286,2016,23,36480,1779,
4,2048-4096,5,64.7111,0.558772,0.177066,20384,315,36480,1779,
5,16-32,5,1.18519,0.00362845,0.0605653,160,135,44096,2229,
5,64-128,5,1.24272,0.0145138,0.231045,640,515,44096,2229,
5,128-256,15,5.14548,0.0761974,0.292956,3360,653,44096,2229,
5,256-512,10,10.3226,0.0725689,0.139076,3200,310,44096,2229,
5,512-1024,10,51.2,0.232221,0.0897263,10240,200,44096,2229,
5,1024-2048,1,87.6522,0.0457184,0.0103185,2016,23,44096,2229,
5,2048-4096,6,62.2901,0.555152,0.176312,24480,393,44096,2229,
6,16-32,5,1.18519,0.00362845,0.0605653,160,135,44096,2229,
6,64-128,5,1.24272,0.0145138,0.231045,640,515,44096,2229,
6,128-256,15,5.14548,0.0761974,0.292956,3360,653,44096,2229,
6,256-512,10,10.3226,0.0725689,0.139076,3200,310,44096,2229,
6,512-1024,10,51.2,0.232221,0.0897263,10240,200,44096,2229,
6,1024-2048,1,87.6522,0.0457184,0.0103185,2016,23,44096,2229,
6,2048-4096,6,62.2901,0.555152,0.176312,24480,393,44096,2229,
