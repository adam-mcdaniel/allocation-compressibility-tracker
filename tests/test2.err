./run.sh: line 15: 793415 Floating point exceptionLD_PRELOAD=$SCRIPT_DIR/libbkmalloc.so BKMALLOC_OPTS="--hooks-file=$SCRIPT_DIR/hook.so --log-hooks" "$PROGRAM_TO_RUN" "$@" 0>&-
