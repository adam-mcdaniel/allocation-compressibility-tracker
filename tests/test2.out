[WARN] Clearing soft dirty bits
[INFO] Elapsed time: 102499 ms
[INFO] Total overhead: 0 ms
[INFO]   Update overhead:     0 ms
[INFO]   Invalidate overhead: 0 ms
[INFO] Malloc count: 4
[INFO] Free count: 1
[INFO] Mmap count: 0
[INFO] Munmap count: 0
[INFO] Total allocations: 4
[INFO] Total frees: 1
[INFO] Heart beat (1 second has elapsed)
[INFO] Interval timer is at 102576ms, test is ready!
[INFO] Running interval
[INFO] Running interval for test 0: Page Tracking Test
[INFO] Interval 1 page tracking starting...
[INFO] Site 0x561D168F72A8 complete, tracked 0 pages
[INFO] Unprotecting 7FF4A2180000, size: 4096
[INFO] Getting pages for allocation at 0x7FF4A2180200
[INFO] Protecting 7FF4A2180000, size: 4096
[INFO] Tracked 1 pages for allocation
[INFO] Unprotecting 7FF4A2180000, size: 4096
[INFO] Unprotecting 7FF4B2C80000, size: 4096
[INFO] Getting pages for allocation at 0x7FF4B2C80100
[INFO] Protecting 7FF4B2C80000, size: 4096
[INFO] Tracked 1 pages for allocation
[WARN] Working thread, giving back access: 0x7FF4B2C8020C
[INFO] Site 0x7FF4B3C96D17 complete, tracked 2 pages
[INFO] Ticking age for 2 pages
